// @progress/kendo-charts/e2e/default-theme.js
import { baseTheme } from '../src/chart/base-theme';
import { $ } from 'e2e-utils';

const DEFAULT_THEME = $.extend(true, {}, baseTheme(), {
    title: {
        color: "#8e8e8e"
    },
    legend: {
        labels: {
            color: "#232323"
        },
        inactiveItems: {
            labels: {
                color: "#919191"
            },
            markers: {
                color: "#919191"
            }
        }
    },
    seriesDefaults: {
        labels: {
            color: "#000",
            background: "#fff",
            opacity: 0.5
        },
        errorBars: {
            color: "#232323"
        },
        candlestick: {
            downColor: "#dedede",
            line: {
                color: "#8d8d8d"
            }
        },
        waterfall: {
            line: {
                color: "#8e8e8e"
            }
        },
        horizontalWaterfall: {
            line: {
                color: "#8e8e8e"
            }
        },
        notes: {
            icon: {
                background: "transparent",
                border: {
                    color: "#8e8e8e"
                }
            },
            label: {
                color: "#232323"
            },
            line: {
                color: "#8e8e8e"
            }
        }
    },
    seriesColors: ["#ff6800", "#a0a700", "#ff8d00", "#678900", "#ffb53c", "#396000"],
    axisDefaults: {
        line: {
            color: "#8e8e8e"
        },
        labels: {
            color: "#232323"
        },
        minorGridLines: {
            color: "#f0f0f0"
        },
        majorGridLines: {
            color: "#dfdfdf"
        },
        title: {
            color: "#232323"
        },
        crosshair: {
            color: "#8e8e8e"
        },
        notes: {
            icon: {
                background: "transparent",
                border: {
                    color: "#8e8e8e"
                }
            },
            label: {
                color: "#232323"
            },
            line: {
                color: "#8e8e8e"
            }
        }
    }
});

const defaults = DEFAULT_THEME.seriesDefaults;
defaults.verticalLine = $.extend({}, defaults.line);
defaults.verticalArea = $.extend({}, defaults.area);
defaults.verticalBoxPlot = $.extend({}, defaults.boxPlot);

defaults.polarArea = $.extend({}, defaults.radarArea);
defaults.polarLine = $.extend({}, defaults.radarLine);

export default DEFAULT_THEME;
