// @progress/kendo-charts/e2e/test-helpers.js
import Chart from '../src/chart/chart';
import DEFAULT_THEME from './default-theme';
import { $ } from 'e2e-utils';

export function createChart(options, themeOptions = DEFAULT_THEME) {
    const element = $("<div id='container' />").appendTo(document.body);

    const chart = new Chart(element[0], options, themeOptions);

    return chart;
}

export function destroyChart(chart) {
    chart.destroy();
    $(chart.element).remove();
}

export function clickChart(chart, element, x = 0, y = 0, pageCoordinates = true) {
    triggerEvent("mousedown", element, x, y, pageCoordinates);
    triggerEvent("mouseup", element, x, y, pageCoordinates);
}

export function hoverChart(chart, element, x = 0, y = 0) {
    chart._mousemove({ pageX: x, pageY: y, target: element });
}

export function triggerEvent(eventName, element, x = 0, y = 0, pageCoordinates = false) {
    const position = pageCoordinates ? { left: 0, top: 0 } : $(element).position();
    const e = new MouseEvent(eventName, {
        bubbles: true,
        clientX: position.left + x,
        clientY: position.top + y });

    element.dispatchEvent(e);
}

export function getChartDomElement(chartElement) {
    return chartElement.visual._observers[0].element;
}
